### DISCUSSION TITLE
[A-L] (5 marzo 2024) Esercizio 3

### DISCUSSION THREAD

**Discussion by IonutCicio** [sapienzastudentsnetwork/progettazione-di-algoritmi#14] [Esercitazioni A-L]
(I. Salvo). Si consideri il grafo diretto $G$ illustrato nella figura qui sotto e l’albero $T$ formato formato dagli archi evidenziati. L’albero $T$ puè essere prodotto da una ricerca in profondità?
- In caso positivo, esibire una rappresentazione di $G$ tramite liste di adiacenza in grado di produrre $T$ e specificare il nodo da cui parte la ricerca e il tipo degli archi ottenuto a seguito della visita.
- In caso negativo, rimpiazzare un arco di $T$ con un altro arco in maniera da ottenere un albero $T'$ con la proprietà che $T'$ possa essere un albero di ricerca per una ricerca in profondità. In tal caso, esibire una rappresentazione di $G$ tramite liste di adiacenza in grado di produrre $T'$ e specificare il nodo da cui parte la visita e il tipo degli archi ottenuto a seguito della visita.

![image](https://github.com/sapienzastudentsnetwork/progalg2324/assets/40034429/1ccd24eb-cd3d-4621-99f9-9ba13262377c)

In fine, che succede se si considera lo stesso grafo $G$ dove però gli archi non sono diretti?

**Comment by Frisbon** [comment]
<html>
<body>
<p>No, non può essere prodotto a partire da nessun nodo (nemmeno $d$) perchè gli archi consentono di “accorciare” il percorso.</p>
<p>Sostituendo ($c,b$) con ($b,c$), il proseguimento ideale del DFS, a partire da $d$, è:</p>
<pre><code class="language-python">d
	f
		h
		g
	c
		a
	e
		b
</code></pre>
<p>La rappresentazione di G con le liste di adiacenza è:</p>

Nodo | Lista
-- | --
a | c
b | a,d,c
c | a
d | c,e,f
e | b,f
f | g,h
g | d,h
h | empty


<p>Se si considera il grafo senza archi orientati, non succede una minchia (te pare che riesci a ricostruire $T'$ con il DFS se hai tutto quel casino?)</p>

</body>
</html>
